run.c

```diff
@@ -371,15 +377,79 @@ long time_in_ms() {
     return time.tv_sec * 1000 + time.tv_nsec / 1000000;
 }
 
+int bpe_lookup(char *str, Piece *vocab, int vocab_size)
+{
+    // lookup 'str' in vocabulary as exact piece match
+    for(int i=0; i<vocab_size; i++) {
+        if( strcmp(str, vocab[i].piece) == 0 ) {
+            return i;
+        }
+    }
+    return -1; // -1 if not found in vocabulary
+}
+
+int bpe_encode(char *text, Piece *vocab, int vocab_size, int *tokens, int *n_tokens)
+{
+    // encode text using vocabulary
+    char merged_str[128]; // max string size in vocabulary * 2
+
+    // start from char sequence
+    *n_tokens = 0;
+    for(char *c=text; *c!='\0'; c++) {
+        sprintf(merged_str, "%c", *c);
+        int id = bpe_lookup(merged_str, vocab, vocab_size);
+        if(id==-1) {
+            return 0;
+        }
+        tokens[*n_tokens] = id;
+        (*n_tokens)++;
+    }
+
+    // merge best consecutive pairs on each iteration
+    while(1) {
+        float best_score = -1e10;
```

Comment:
did you see that some of the scores in sentencepieve for some vocab elements is -1e9? What's up with that :S

Reply:
well, I've seen negative scores and -1e10 was just a guess. But looks like lucky one?

```
In [22]: s = spm.SentencePieceProcessor(model_file='tokenizer.model')

In [23]: min([s.get_score(i) for i in range(32000)])
Out[23]: -1000000000.0
```

Reply:
I guess I don't really understand what the -1e10 tokens are? Why are some tokens at random score -1e10

Reply:
No idea about scoring mechanism but these are tokens with -1e9 score in the model:
```
In [29]: [s.id_to_piece(i) for i in range(32000) if s.get_score(i)==-1e9]
Out[29]: 
['▁▁',
 '▁▁▁▁',
 '▁▁▁▁▁▁▁▁',
 '▁▁▁▁▁',
 '▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁',
 '▁▁▁▁▁▁',
 '▁▁▁▁▁▁▁▁▁▁▁▁',
 '▁▁▁▁▁▁▁▁▁▁▁▁▁',
 '▁▁▁▁▁▁▁▁▁▁',
 '▁▁▁▁▁▁▁▁▁▁▁▁▁▁',
 '▁▁▁',
 '▁▁▁▁▁▁▁▁▁',
 '▁▁▁▁▁▁▁',
 '▁▁▁▁▁▁▁▁▁▁▁',
 '▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁',
 '▁']
```